framework/src/eventloop/eventloop_timer.c

```diff
@@ -0,0 +1,172 @@
+ *
+ ****************************************************************************/
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <stdio.h>
+#include <libtuv/uv.h>
+#include <eventloop/eventloop.h>
+
+#include "eventloop_internal.h"
+
+typedef struct el_timer_s {
+	timeout_callback func;
+	void *cb_data;
+	event_timer *handle;
+	struct el_timer_s *next;
+} el_timer_t;
+
+el_timer_t *g_timers = NULL;
+
+static void add_timer_list(el_timer_t *timer)
+{
+	el_timer_t *iter;
+
+	if (g_timers == NULL) {
+		g_timers = timer;
+		return;
+	}
+
+	iter = g_timers;
+
+	/* Add a new timer to the end of timer list */
+	while (iter->next) {
+		iter = iter->next;
+	}
+	iter->next = timer;
+}
+
+static void remove_timer_list(event_timer *handle)
+{
+	el_timer_t *iter;
+	el_timer_t *next_iter;
+
+	if (g_timers == NULL) {
+		return;
+	}
+
+	iter = g_timers;
+
+	/* Remove the first timer from the timer list */
+	if (iter->handle == handle) {
+		g_timers = iter->next;
+		free(handle);
+		free(iter);
+		return;
+	}
+
+	/* Search and Remove the timer from timer list */
+	while (iter->next) {
+		next_iter = iter->next;
+		if (next_iter->handle == handle) {
+			iter->next = next_iter->next;
+			free(handle);
+			free(next_iter);
+			break;
+		}
+		iter = next_iter;
+	}
+}
+
+static void timeout_callback_func(event_timer *handle)
+{
+	el_timer_t *timer;
+
+	if (!handle || !handle->data) {
+		return;
+	}
+
+	timer = handle->data;
+
+	if (timer->func) {
+		timer->func(timer->cb_data);
+	}
+
+	uv_timer_stop(timer->handle);
+	free(timer->handle);
+	free(timer);
+}
+
+int eventloop_add_timeout(unsigned int timeout, timeout_callback func, void *data)
+{
+	el_timer_t *timer = NULL;
+
+	timer = (el_timer_t *)malloc(sizeof(el_timer_t));
+	if (timer == NULL) {
+		return ERROR;
+	}
+	timer->func = func;
+	timer->cb_data = data;
+	timer->handle = (event_timer *)malloc(sizeof(event_timer));
+	if (timer->handle == NULL) {
+		free(timer);
+		return ERROR;
+	}
+	timer->handle->data = timer;
+	timer->next = NULL;
+
+	uv_update_time(get_loop());
+	uv_timer_init(get_loop(), timer->handle);
+	uv_timer_start(timer->handle, timeout_callback_func, timeout, 0);
+
+	return OK;
+}
+
+event_timer *eventloop_add_timer(unsigned int timeout, unsigned int interval, timeout_callback func, void *data)
+{
+	el_timer_t *timer = NULL;
+
+	timer = (el_timer_t *)malloc(sizeof(el_timer_t));
+	if (timer == NULL) {
+		return NULL;
+	}
+	timer->func = func;
+	timer->cb_data = data;
+	timer->handle = (el_timer_t *)malloc(sizeof(event_timer));
+	if (timer->handle == NULL) {
+		free(timer);
+		return NULL;
+	}
+	timer->handle->data = timer;
+	timer->next = NULL;
+
+	/* Add a timer to the list of timers, g_timers */
+	add_timer_list(timer);
+
+	uv_update_time(get_loop());
+	uv_timer_init(get_loop(), timer->handle);
```

Comment:
It follows calling step.
init->start, right?
```
if (init()) {
   start();
}
```
Probably Above will be work properly

Reply:
Yes. but uv_timer_init always returns 0.